Sources/NIOCore/Pool.swift

```diff
@@ -0,0 +1,52 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+public class Pool<Element> {
+    private let maxSize: Int
+    private let ctor: () -> Element
+    private let dtor: (Element) -> Void
+    private var elements: [Element]
+
+    public init(maxSize: Int, ctor: @escaping () -> Element, dtor: @escaping (Element) -> Void) {
```

Comment:
Rather than using separate constructor/destructor functions, we'd typically want to define a protocol to which `Element` conforms that implements the relevant methods. Something like:

```
protocol PoolElement {
    init()

    func evictedFromPool()
}

Reply:
I had some debates regarding such approach which is I see is common for swift, but sometimes it does not work well from my point of view. So we are going to extend some type with 'PoolElement' protocol. But is it really needed? Why the particular type should expose the 'PoolElement' functionality and why it should even know that it is going to be pooled with some pool? Often it is only the particular pool who knows what does it going to pool and how...
As you can see in the PR there are 2 pool
- pool to be used in the event loop, which just allocates buffer
- pool to be used in tests, which allocates buffers with markers and check those markers when buffer is not needed any more...
The PoolElement protocol does not give a possibility to have a different pooling logic for the same type in different pools... or at least I can't see a simple way for that...
Or am I missing something?

Reply:
The thing you're missing is that the way to pool the "same thing" in different ways is to define a new type with the new behaviour. Swift `struct`s are essentially "free": they're trivial aggregates. That means whenever we want a different behaviour we can wrap the type we want in a new structure that adds that new behaviour.


In general if we were defining a public API I'd be a bit more cautious, but as this is purely an internal data structure I think it's probably nicer to avoid unstructured callbacks.

Reply:
So, I made another branch with a Pool<> which operates with something implementing 'PoolElement'.
I did a PoolableBuffer struct implementing default behaviour: https://github.com/ordo-one/swift-nio/blob/aa23e675eb95292f98c08391cf30bec37fed8292/Sources/NIOPosix/SelectableEventLoop.swift#L33
My understanding was we can define a different behaviour for a poolable buffer adding some bounds checks, like I did at https://github.com/ordo-one/swift-nio/blob/aa23e675eb95292f98c08391cf30bec37fed8292/Tests/NIOPosixTests/ChannelTests.swift#L76, but TestPooledBuffer can not be used as a pooled type, the code does not compile...
Am I missing something or did I get you wrong?

Reply:
When you say the code doesn't compile, what did you try to do and what happened?

Reply:
Well, my understanding was you want to have 2 different types to be used as an argument type for the pool.
1) PooledBuffer, with a common allocation logic
2) TestPooledBuffer with allocation and bounds check
and depending on the case we could use pools with different types.
But the problem here we can not keep something like Pool<PoolElement> in the PendingWritesManager, because compiler (and I understand it) want to know a particular type here.
So, we should specify a particular at https://github.com/ordo-one/swift-nio/blob/pool-iovecs-buffers2/Sources/NIOPosix/SelectableEventLoop.swift#L125
And after that we should always use Pool<PooledBuffer> as an argument for the PendingWritesManager initialiser, and can't provide something different, like for example Pool<TestPooledBuffer>....

Reply:
To be clear, I agree with this: I think the pooled buffer should have bounds checks in debug mode, but that's a different flow. I'm not proposing we use a protocol so that we can support multiple different types in the core code, just so that we can provide a nicer internal abstraction for ourselves.

Reply:
Then I am totally confused...
In a very first message in that thread you suggest to use a protocol for the pool element.
Not you tell you are not proposing to use protocol.
So what should we do now?

Reply:
Sorry, that might not have been clear. I _am_ proposing a protocol. When I said "I'm not proposing a protocol", that was part of a sentence that said "I'm not proposing it because X, but because Y".

Reply:
So could you give some hints on how can have a protocol, supposed to be implemented by types pooled in the Pool'<>', and have 2 different implementations of that protocol for prod and tests?

Reply:
Do I understand correctly that only a single type is suggested but with assertions for the invariants etc that simply won’t be part of optimized builds?

Reply:
Yes, exactly so.

Reply:
Yes, much closer. The extra thing to add is a bunch of `assert`s in the `get` function to ensure that the bounds are respected there too.

Reply:
Do you mean PooledBuffer.get()?

Reply:
I do.

Reply:
What kind of assertions do you think off?

Reply:
Confirming that the start addresses and end addresses don't overlap, and that they fit within the buffer.

Reply:
I pushed a commit with some bounds checks.
https://github.com/apple/swift-nio/commit/a437a52ed134c90f30152e8811c59fe0cb51f295
Could you have a look?

Reply:
Looks great!

Reply:
Do you think we can merge these changes?
What will be better, create a new PR from that branch, or apply those changes to the branch where that PR is made from?

Reply:
I think you can merge those changes into this branch.